### PR TITLE
[Zoe]: Part 1 of Switching to Zoe Invites

### PR DESCRIPTION
**Pull request by katelynsills**
This PR is Part 1 of the switch to using Zoe invites rather than escrow receipts. It includes changes to:

* state.js
* zoe.js
* zoe.chainmail
* various 'helper' files which are deleted or changed, such as `seatConfig` and `escrowReceiptConfig`
* automaticRefund contract and tests

It DOES NOT include changes to the other contracts. I have SKIPPED those tests and will be making separate PRs per contract.

### CODE REVIEW

File: packages/zoe/zoe.js [hunk 1]
```diff
@@ -9,31 +9,28 @@ import { isOfferSafeForAll } from './isOfferSafe';
...
-
-  const escrowReceiptMint = makeMint(
-    'zoeEscrowReceipts',
-    makeEscrowReceiptConfig,
-  );
-  const escrowReceiptAssay = escrowReceiptMint.getAssay();
+  const makeInviteMint = () => {
+    const handleToSeat = new WeakMap();
+    const inviteMint = makeMint('zoeInvite', makeInviteConfig);
+    return harden({
```
Comment (erights) [inline packages/zoe/zoe.js]: I was confused by this about three times reading the code. The function on line 20 is called `makeInviteMint`. The variable on line 22 is called `inviteMint`. But the object named by that variable is not what a call to `makeInviteMint` returns. The object that is returned does not have the API described by mint.chainmail, and so is not even a Mint.

File: packages/zoe/zoe.js [hunk 2]
```diff
@@ -9,31 +9,28 @@ import { isOfferSafeForAll } from './isOfferSafe';
...
+      inviteAssay: inviteMint.getAssay(),
+      mintInvite: (handle, seat, units) => {
+        handleToSeat.set(handle, seat);
+        const purse = inviteMint.mint(units);
+        return purse.withdrawAll();
+      },
+      getInviteSeat: handle => handleToSeat.get(handle),
+    });
+  };
+  const { inviteAssay, mintInvite, getInviteSeat } = makeInviteMint();
```
Comment (erights) [inline packages/zoe/zoe.js]: If I'm reading this right, `inviteAssay` is a function that always returns the same assay. Why not just bind it directly to the assay?

In fact, it looks like `makeInviteMint` is only used here, and the object it returns is only used to initialize these three variables. Why not just unpack the code of `makeInviteMint` into code that initializes these three variables directly?

File: packages/zoe/zoe.js [hunk 3]
```diff
@@ -9,31 +9,28 @@ import { isOfferSafeForAll } from './isOfferSafe';
...
-    addUseObj: addInviteSeat,
-  } = makeSeatMint('zoeInvite');
-
-  const escrowReceiptMint = makeMint(
-    'zoeEscrowReceipts',
-    makeEscrowReceiptConfig,
-  );
-  const escrowReceiptAssay = escrowReceiptMint.getAssay();
+  const makeInviteMint = () => {
+    const handleToSeat = new WeakMap();
```
Comment (erights) [inline packages/zoe/zoe.js]: Why a `WeakMap` rather than a `PrivateName`?

**Review by erights — CHANGES_REQUESTED** [review]
Review so far. Will get back to it this afternoon. Could you fix these before then? Thanks.

File: packages/zoe/zoe.js (hunk 3 above)
Comment (katelynsills) [reply packages/zoe/zoe.js]: No particular reason. This code was ported from `seatMint.js` line 12, not written recently. I'll change it to a `PrivateName`.

File: packages/zoe/zoe.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/zoe.js]: inviteAssay is bound directly to the assay: `inviteAssay: inviteMint.getAssay(),`

I wanted makeInviteMint to be a separate function for POLA. `handleToSeat` isn't accessible outside the function, and the only access is through `mintInvite` and `getInviteSeat`.

File: packages/zoe/inviteConfig.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { insist } from '@agoric/ertp/util/insist';
+
+const insistInviteExtent = inviteExtent => {
+  const properties = Object.getOwnPropertyNames(inviteExtent);
+  // The `handle` is how the use object will be looked up
+  insist(properties.includes('handle'))`must include 'handle'`;
+  insist(
+    passStyleOf(inviteExtent.handle) === 'presence' &&
+      Object.entries(inviteExtent.handle).length === 0 &&
+      inviteExtent.handle.constructor === Object,
```
[the remark this answers is not in the document]
Comment (katelynsills) [reply packages/zoe/inviteConfig.js]: How do I test for that? Would it be:

```
Object.getPrototypeOf(inviteExtent.handle) === Object.prototype;
```

File: packages/zoe/inviteConfig.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  )`handle should be an empty object`;
+  insist(
```
[the remark this answers is not in the document]
Comment (katelynsills) [reply packages/zoe/inviteConfig.js:21]: I would like to, but the problem is that apart from `handle` and `instanceHandle`, the rest of the inviteExtent is up to the contracts so I didn't think I was able to test further. I can add a check for `instanceHandle`. If there's more to check for, please let me know.

File: packages/zoe/zoe.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/zoe.js]: Ok, it seems like the name of the function is what is creating the wrong expectation. I'll change the name of the function. I think other than that, it's pretty straightforward. I'm encapsulating all of the invite logic so all that Zoe needs to know is that there is an `inviteAssay`, and two functions, `mintInvite` and `getInviteSeat`. I think the encapsulation is valuable. Zoe shouldn't have to know about `handleToSeat`.

File: packages/zoe/inviteConfig.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (katelynsills) [reply packages/zoe/inviteConfig.js]: ^ I added the above check - let me know if there's a better one

File: packages/zoe/zoe.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/zoe.js]: I just pushed up a commit that exposes `handleToSeat` and unpacks the code. I still don't like exposed weakMaps, but it is simpler to understand what is happening here.

File: packages/zoe/zoe.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/zoe.js]: I said this in a different comment thread, but I ended up taking this code out of the function. Let me know if that works better for you!

File: packages/zoe/inviteConfig.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (erights) [reply packages/zoe/inviteConfig.js]: That one is fine. `passStyleOf` uses the same test. Resolving.

File: packages/zoe/zoe.js (hunk 1 above)
Comment (erights) [reply packages/zoe/zoe.js]: I like the way it reads now. Addresses all my concerns here. Resolving.

File: packages/zoe/zoe.js (hunk 2 above)
Comment (erights) [reply packages/zoe/zoe.js]: I like the new code better than the old code here. But I appreciate the POLA concern. We can revisit that sometime if you like, but let's stick with your new code for now.

Comment (katelynsills) [reply packages/zoe/zoe.js]: Sounds good!

**Review by erights — APPROVED** [review]
LGTM!!

File: packages/zoe/zoe.chainmail [hunk 6]
```diff
@@ -81,14 +67,14 @@ interface Zoe {
    * payoutRules. If the payoutRules kind is 'offerAtMost',
```
[the remark this answers is not in the document]
Comment (katelynsills) [reply packages/zoe/zoe.chainmail:67]: Good catch, I'll make an issue of this

File: packages/zoe/zoe.chainmail [hunk 7]
```diff
@@ -110,16 +96,24 @@ struct ExitRule ( ) {
   deadline :Number;
 }
```
[the remark this answers is not in the document]
Comment (katelynsills) [reply packages/zoe/zoe.chainmail:97]: Ah, good point

**Pull request merged**